hpx/lcos/local/detail/condition_variable.hpp

```diff
@@ -72,280 +93,78 @@ namespace hpx { namespace lcos { namespace local { namespace detail
-            HPX_ASSERT_OWNS_LOCK(lock);
-
-            return queue_.empty();
-        }
-
-        template <typename Mutex>
-        std::size_t size(boost::unique_lock<Mutex> const& lock) const
-        {
-            HPX_ASSERT_OWNS_LOCK(lock);
+        HPX_EXPORT bool empty(
+            boost::unique_lock<mutex_type> const& lock) const;
 
-            return queue_.size();
-        }
+        HPX_EXPORT std::size_t size(
+            boost::unique_lock<mutex_type> const& lock) const;
 
         // Return false if no more threads are waiting (returns true if queue
         // is non-empty).
-        template <typename Mutex>
-        bool notify_one(boost::unique_lock<Mutex> lock, error_code& ec = throws)
-        {
-            HPX_ASSERT_OWNS_LOCK(lock);
+        HPX_EXPORT bool notify_one(
+            boost::unique_lock<mutex_type> lock, error_code& ec = throws);
 
-            if (!queue_.empty())
-            {
-                threads::thread_id_repr_type id = queue_.front().id_;
+        HPX_EXPORT void notify_all(
+            boost::unique_lock<mutex_type> lock, error_code& ec = throws);
 
-                // remove item from queue before error handling
-                queue_.front().id_ = threads::invalid_thread_id_repr;
-                queue_.pop_front();
+        HPX_EXPORT void abort_all(
+            boost::unique_lock<mutex_type> lock);
 
-                if (HPX_UNLIKELY(id == threads::invalid_thread_id_repr))
-                {
-                    lock.unlock();
+        HPX_EXPORT threads::thread_state_ex_enum wait(
+            boost::unique_lock<mutex_type>&& lock,
+            char const* description, error_code& ec = throws);
 
-                    HPX_THROWS_IF(ec, null_thread_id,
-                        "condition_variable::notify_one",
-                        "NULL thread id encountered");
-                    return false;
-                }
-
-                bool not_empty = !queue_.empty();
-                lock.unlock();
-
-                threads::set_thread_state(threads::thread_id_type(
-                        reinterpret_cast<threads::thread_data*>(id)),
-                    threads::pending, threads::wait_signaled,
-                    threads::thread_priority_default, ec);
-                return not_empty;
-            }
-
-            if (&ec != &throws)
-                ec = make_success_code();
-
-            return false;
-        }
-
-        // re-add the remaining items to the original queue
-        template <typename Mutex>
-        void prepend_entries(boost::unique_lock<Mutex>& lock, queue_type& queue)
+        threads::thread_state_ex_enum wait(
+            boost::unique_lock<mutex_type>& lock,
+            char const* description, error_code& ec = throws)
         {
-            HPX_ASSERT_OWNS_LOCK(lock);
-
-            // splice is constant time only if it == end
-            queue.splice(queue.end(), queue_);
-            queue_.swap(queue);
+            relock_guard rl(lock);
```

Comment:
Since the lock is moved into wait down below, is it really save to relock it again afterwards?